rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,221 @@
+
+In the most common mode of operation, Database Access connects the user to the database as a requested user, but does not verify if the user exists or has the appropriate rights. This task is left for the database administrator to perform. Automatic user creation is a relatively new Database Access feature that automates part of this task in supported database configurations. On connection, a new database user is created (or an old one reactivated) with a name matching that of the Teleport user. This new user is given roles from a fixed list of roles. Teleport expects the roles to be managed manually and has no insight into their definition.
+
+This RFD introduces a mechanism for _database permission management_, a way to define the effective database user permissions within Teleport roles and automatically apply them as needed. This provides visibility into the effective permissions and eases the burden on database administrators.
+
+## Details
+
+### UX
+
+#### Usage
+
+Teleport administrators modify the role definition assigned to the user, configuring both automatic user creation and individual permissions for the database.
+
+On connection, the engine calculates the effective permissions from the roles and database schema.
+
+After automatically creating the database user for the Teleport user who initiated the connection, the calculated set of permissions is applied to the database user.
+
+After the session is finished, the user is removed/deactivated, and all permissions are revoked.
+
+#### Configuration
+
+1. User roles will be extended with a new `db_permissions` field in both `allow` and `deny` parts of the spec.
+2. The new field will contain a list of DB permissions to be applied for the user.
+3. Each individual DB permission will consist of four fields:
+
+   ```protobuf
+   // DatabasePermission is the single permission to be granted.
+   message DatabasePermission {
+   // Version is the version of permission. Combined with protocol it dictates the meaning of other fields.
+   string Version = 1 [(gogoproto.jsontag) = "version"];
+   // Protocol is the database protocol of permission.
+   string Protocol = 2 [(gogoproto.jsontag) = "protocol"];
+   // Target is the target of the permission, for example the name of database table.
+   string Target = 3 [(gogoproto.jsontag) = "target"];
+   // Access is the level of access granted by this permission to target.
+   string Access = 4 [(gogoproto.jsontag) = "access"];
+   }
+   ```
+
+   An example instance:
+
+   ```yaml
+   spec:
+     allow:
+       db_permissions:
+         - protocol: postgres
+           version: v1
+           target: /public/customers
+           access: SELECT
```

Comment:
TBH I'm not a huge fan of this structure:

- Having to specify protocol in the role is very confusing, I don't think roles need to have protocols.
- I agree with Marek that version also seems out of place, we don't have a concept of sub-resources, so if you need to version something it needs to be a proper separate resource.
- Target seems like is a loosely-formatted (and probably database type specific) string that doesn't make it clear what is actually being targeted.

Previously, Marek and I discussed a label-based approach. We would introspect the database schema, allow users to create "rules" to mark database objects (like individual databases, tables, etc.) with labels and then allow/deny access in the roles using these labels. This would result in a much less coupled structure and easier to reason about design.

Can we explore this approach in the RFD?

Reply:
Sure, I'll try this.

Reply:
Thinking about this further, I'm not quite sure how to approach this bit:

> We would introspect the database schema, allow users to create "rules" to mark database objects (like individual databases, tables, etc.) with labels and then allow/deny access in the roles using these labels.

Can you elaborate on this @r0mant?

Reply:
Sure, so the inspiration partially comes from our existing [Okta integration](https://goteleport.com/docs/application-access/okta/architecture/) in which we import all users and groups from Okta and operators can configure [import rules](https://goteleport.com/docs/application-access/okta/architecture/#okta-import-rules) (as Teleport resources) that allow to apply labels to imported objects.

We could have, say, a database import rule that would similarly allow to apply labels to imported database objects (just spitballing, exact syntax could use some thought):

```yaml
kind: db_import_rule
version: v1
metadata:
  name: test-rule
spec:
  priority: 10
  mappings:
  - match:
    - db_service: postgres
      schema: postgres
      table: users
    add_labels:
      critical: yes
```

Which would result in a resource existing in Teleport representing a database table, e.g.:

```yaml
kind: db_table
version: v1
metadata:
  name: users
  labels:
    critical: yes
spec:
  schema: postgres
```

Then, a user could have a role that grants them, say, access to read this table, something like:

```yaml
kind: role
version: v7
metadata:
  name: example
spec:
  allow:
    db_tables:
      read:
        critical: yes
```

Like I said the exact syntax could use some more thought and how all these pieces fit together is what I'd like us to explore in this RFD but hopefully the general idea is clear now?

Reply:
I've updated the design to be more aligned with this idea, please check it out.